File: ruby/lib/mable_bank/application/transfer_journal.rb
# frozen_string_literal: true

module MableBank
  module Application
    TransferJournalEntry = Data.define(
      :run_id,
      :transfer_id,
      :outcome,
      :reason_code,
      :instruction,
      :recorded_at
    )

    class TransferJournal
      def record(_entry)
        raise NotImplementedError
      end

      def entries
        raise NotImplementedError
      end
    end

    class InMemoryTransferJournal < TransferJournal
      attr_reader :entries

      def initialize
        super
        @entries = []
      end

      def record(entry)
        @entries << entry
      end
    end
  end
end

File: nodejs-typescript/src/application/TransferJournal.ts
import type { TransferId } from '../domain/TransferId.js';
import type { TransferInstruction } from '../domain/TransferInstruction.js';

export type TransferJournalEntry = {
  readonly runId: string;
  readonly transferId: TransferId;
  readonly outcome: 'succeeded' | 'failed' | 'skipped';
  readonly reasonCode: string | undefined;
  readonly instruction: TransferInstruction;
  readonly recordedAt: Date;
};

export interface TransferJournal {
  readonly entries: readonly TransferJournalEntry[];
  record(entry: TransferJournalEntry): void;
}

export class InMemoryTransferJournal implements TransferJournal {
  private readonly _entries: TransferJournalEntry[] = [];

  get entries(): readonly TransferJournalEntry[] {
    return this._entries;
  }

  record(entry: TransferJournalEntry): void {
    this._entries.push(entry);
  }
}
